src/attached_probe.cpp

```diff
@@ -811,48 +811,56 @@ void AttachedProbe::load_prog(BPFfeature &feature)
         if (bt_debug == DebugLevel::kNone)
           silencer.silence();
 
-        LIBBPF_OPTS(bpf_btf_load_opts,
-                    btf_opts,
-                    .log_buf = log_buf.get(),
-                    .log_level = static_cast<__u32>(log_level),
-                    .log_size = static_cast<__u32>(log_buf_size), );
-
-        auto &btf = prog_.getBTF();
-        auto btf_fd = bpf_btf_load(btf.data(), btf.size(), &btf_opts);
-
-        opts.prog_btf_fd = btf_fd;
+        // Not all older kernels cope well with BTF loaded (see #2913):
+        // On the 3rd attempt, retry bpf_prog_load() without loading BTF
+        int btf_fd = -1;
+        if (attempt < 2) {
+          LIBBPF_OPTS(bpf_btf_load_opts,
+                      btf_opts,
+                      .log_buf = log_buf.get(),
+                      .log_level = static_cast<__u32>(log_level),
+                      .log_size = static_cast<__u32>(log_buf_size), );
+
+          auto &btf = prog_.getBTF();
+          btf_fd = bpf_btf_load(btf.data(), btf.size(), &btf_opts);
+          if (btf_fd >= 0)
+            opts.prog_btf_fd = (__u32)btf_fd;
+        }
 
-        if (!func_infos.empty()) {
+        // Attempt .BTF.ext func_info relocation, but older kernels
+        // do not cope well with func_infos. See #2913 for reference:
+        // So, on retries, call bpf_prog_load() without func_infos:
+        if (attempt == 0 && !func_infos.empty()) {
           opts.func_info_rec_size = sizeof(struct bpf_func_info);
           opts.func_info = func_infos.data();
           opts.func_info_cnt = func_infos.size() / sizeof(struct bpf_func_info);
         }
 
-        // Don't attempt to load the program if the BTF load failed.
-        // This will fall back to the error handling for failed program load,
-        // which is more robust.
-        if (btf_fd >= 0) {
-          progfd_ = bpf_prog_load(static_cast<::bpf_prog_type>(prog_type),
-                                  name.c_str(),
-                                  license,
-                                  reinterpret_cast<const struct bpf_insn *>(
-                                      insns.data()),
-                                  insns.size() / sizeof(struct bpf_insn),
-                                  &opts);
+        // Most bpftrace programs still work without func_infos and without BTF.
+        // See the comments above for reference(these are the only retries)
```

Comment:
This comment doesn't make much sense here (both BTF and func_info is passed via `opts` so it's not used in the below command). I'd rather see a more general comment somewhere (maybe before the "attempt" loop header?) explaining what features are omitted in which attempt.

Reply:
Yes, that's best, done now with the new outer loop!